File: backend-go/internal/handler/rentals.go
package handler

import (
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/gin-gonic/gin"

	"github.com/instaagrammeta/rentacar/backend-go/internal/service"
)

// parseReturnDate parses an ISO datetime, defaulting to now.
func parseReturnDate(value string) time.Time {
	if value == "" {
		return time.Now()
	}
	if t, err := time.Parse(time.RFC3339, value); err == nil {
		return t
	}
	if t, err := time.Parse("2006-01-02T15:04:05", value); err == nil {
		return t
	}
	if len(value) >= 10 {
		if t, err := time.Parse("2006-01-02", value[:10]); err == nil {
			return t
		}
	}
	return time.Now()
}

// ListRentals returns rentals filtered by status.
func (h *Handler) ListRentals(c *gin.Context) {
	page, perPage, offset := pagination(c)
	status := strings.TrimSpace(c.Query("status"))
	result, err := h.Svc.ListRentals(status, page, perPage, offset)
	if err != nil {
		respondError(c, err)
		return
	}
	c.JSON(http.StatusOK, result)
}

// GetRental returns a single rental including its return record.
func (h *Handler) GetRental(c *gin.Context) {
	id, ok := paramID(c, "id")
	if !ok {
		return
	}
	rental, err := h.Svc.GetRental(id)
	if err != nil {
		respondError(c, err)
		return
	}
	data := rental.ToMap()
	if rental.VehicleReturn != nil {
		data["vehicle_return"] = rental.VehicleReturn.ToMap()
	}
	c.JSON(http.StatusOK, data)
}

// CreateRental creates a rental contract.
func (h *Handler) CreateRental(c *gin.Context) {
	var in service.RentalInput
	_ = c.ShouldBindJSON(&in)
	rental, err := h.Svc.CreateRental(in, actorFrom(c))
	if err != nil {
		respondError(c, err)
		return
	}
	c.JSON(http.StatusCreated, rental.ToMap())
}

// CancelRental cancels a rental.
func (h *Handler) CancelRental(c *gin.Context) {
	id, ok := paramID(c, "id")
	if !ok {
		return
	}
	rental, err := h.Svc.CancelRental(id, actorFrom(c))
	if err != nil {
		respondError(c, err)
		return
	}
	c.JSON(http.StatusOK, rental.ToMap())
}

// DownloadContract serves the rental contract PDF.
func (h *Handler) DownloadContract(c *gin.Context) {
	id, ok := paramID(c, "id")
	if !ok {
		return
	}
	rental, err := h.Svc.GetRental(id)
	if err != nil {
		respondError(c, err)
		return
	}
	if rental.PDFPath == nil || *rental.PDFPath == "" {
		rental, err = h.Svc.RegeneratePDF(id)
		if err != nil {
			respondError(c, err)
			return
		}
	}
	if rental.PDFPath == nil || *rental.PDFPath == "" {
		c.JSON(http.StatusNotFound, gin.H{"error": "Договор не найден"})
		return
	}
	full := h.Svc.Media.AbsoluteExportPath(*rental.PDFPath)
	c.FileAttachment(full, rental.ContractNumber+".pdf")
}

// frontendBaseURL derives the public frontend base URL (scheme://host[:port])
// from the incoming request, so generated QR links point to whatever address
// the staff member is actually using. Falls back to the configured PublicURL.
func frontendBaseURL(c *gin.Context, fallback string) string {
	if o := strings.TrimSpace(c.GetHeader("Origin")); o != "" {
		return strings.TrimRight(o, "/")
	}
	if ref := strings.TrimSpace(c.GetHeader("Referer")); ref != "" {
		if u, err := url.Parse(ref); err == nil && u.Scheme != "" && u.Host != "" {
			return u.Scheme + "://" + u.Host
		}
	}
	return fallback
}

// RentalQR returns (and regenerates) the public QR code for a rental. The QR
// encodes a link to the public status page the client opens to see how much
// time is left. The link host is taken from the request origin so it always
// points to a reachable address.
func (h *Handler) RentalQR(c *gin.Context) {
	id, ok := paramID(c, "id")
	if !ok {
		return
	}
	base := frontendBaseURL(c, h.Cfg.PublicURL)
	rental, err := h.Svc.RegenerateRentalQR(id, base)
	if err != nil {
		respondError(c, err)
		return
	}
	c.JSON(http.StatusOK, gin.H{
		"qr_code_path": rental.QRCodePath,
		"public_token": rental.PublicToken,
		"public_url":   h.Svc.PublicRentalURLWithBase(rental, base),
	})
}

// PublicRental returns the public, unauthenticated view of a rental looked up
// by its opaque token — used by the page a client reaches via the QR code.
func (h *Handler) PublicRental(c *gin.Context) {
	token := c.Param("token")
	rental, err := h.Svc.GetRentalByToken(token)
	if err != nil {
		respondError(c, err)
		return
	}
	c.JSON(http.StatusOK, h.Svc.PublicRentalView(rental))
}

// PreviewReturn calculates return charges without persisting.
func (h *Handler) PreviewReturn(c *gin.Context) {
	id, ok := paramID(c, "id")
	if !ok {
		return
	}
	rental, err := h.Svc.GetRental(id)
	if err != nil {
		respondError(c, err)
		return
	}
	var body struct {
		ReturnDate string  `json:"return_date"`
		DamageCost float64 `json:"damage_cost"`
		Penalties  float64 `json:"penalties"`
	}
	_ = c.ShouldBindJSON(&body)
	rd := parseReturnDate(body.ReturnDate)
	calc := h.Svc.CalculateReturn(rental, rd, body.DamageCost, body.Penalties)
	c.JSON(http.StatusOK, calc.ToMap())
}

// CreateReturn records a vehicle return.
func (h *Handler) CreateReturn(c *gin.Context) {
	id, ok := paramID(c, "id")
	if !ok {
		return
	}
	var in service.ReturnInput
	_ = c.ShouldBindJSON(&in)
	vr, err := h.Svc.CreateReturn(id, in, actorFrom(c))
	if err != nil {
		respondError(c, err)
		return
	}
	c.JSON(http.StatusCreated, vr.ToMap())
}

File: backend/app/api/rentals.py
"""Rental contract & vehicle return endpoints (Modules 6 & 7)."""
from __future__ import annotations

from flask import Blueprint, jsonify, request, send_file

from app.api.helpers import current_user, get_pagination, json_body
from app.models import Rental
from app.models.enums import RentalStatus
from app.repositories import rental_repo
from app.services import rental_service, return_service
from app.utils.security import permission_required

from flask import current_app

bp = Blueprint("rentals", __name__, url_prefix="/api/rentals")


@bp.get("")
@permission_required("rentals")
def list_rentals():
    page, per_page, offset = get_pagination()
    status = request.args.get("status", "").strip()
    filters = []
    if status:
        filters.append(Rental.status == RentalStatus(status))
    total = rental_repo.count(filters or None)
    items = rental_repo.list(filters=filters or None, limit=per_page, offset=offset)
    return jsonify(
        {"items": [r.to_dict() for r in items], "total": total, "page": page, "per_page": per_page}
    )


@bp.get("/<int:rental_id>")
@permission_required("rentals")
def get_rental(rental_id: int):
    rental = rental_repo.get_or_404(rental_id)
    data = rental.to_dict()
    if rental.vehicle_return:
        data["vehicle_return"] = rental.vehicle_return.to_dict()
    return jsonify(data)


@bp.post("")
@permission_required("rentals")
def create_rental():
    rental = rental_service.create_rental(json_body(), actor=current_user())
    return jsonify(rental.to_dict()), 201


@bp.post("/<int:rental_id>/cancel")
@permission_required("rentals")
def cancel_rental(rental_id: int):
    rental = rental_service.cancel_rental(rental_id, actor=current_user())
    return jsonify(rental.to_dict())


@bp.get("/<int:rental_id>/contract")
@permission_required("rentals")
def download_contract(rental_id: int):
    rental = rental_repo.get_or_404(rental_id)
    if not rental.pdf_path:
        rental = rental_service.regenerate_pdf(rental_id)
    full_path = current_app.config["EXPORT_DIR"] / rental.pdf_path
    return send_file(full_path, as_attachment=True, download_name=f"{rental.contract_number}.pdf")


# ----------------------------------------------------------- vehicle returns
@bp.post("/<int:rental_id>/return/preview")
@permission_required("returns")
def preview_return(rental_id: int):
    """Calculate return charges without persisting (live preview in the UI)."""
    from datetime import datetime

    rental = rental_repo.get_or_404(rental_id)
    body = json_body()
    return_date = body.get("return_date")
    parsed = (
        datetime.fromisoformat(return_date) if return_date else datetime.now()
    )
    calc = return_service.calculate_return(
        rental,
        return_date=parsed,
        damage_cost=float(body.get("damage_cost") or 0),
        penalties=float(body.get("penalties") or 0),
    )
    return jsonify(calc)


@bp.post("/<int:rental_id>/return")
@permission_required("returns")
def create_return(rental_id: int):
    vehicle_return = return_service.create_return(rental_id, json_body(), actor=current_user())
    return jsonify(vehicle_return.to_dict()), 201
